Brokerages/Bitfinex/BitfinexBrokerage.cs

```diff
@@ -0,0 +1,449 @@
+{
+
+    /// <summary>
+    /// Bitfinex exchange REST integration.
+    /// </summary>
+    public partial class BitfinexBrokerage : Brokerage, IDataQueueHandler
+    {
+
+        #region Declarations
+        /// <summary>
+        /// Ticks collection
+        /// </summary>
+        protected List<Tick> Ticks = new List<Tick>();
+        CancellationTokenSource _tickerToken;
+        /// <summary>
+        /// Divisor for prices. Scales prices/volumes to allow trades on 0.01 of unit
+        /// </summary>
+        protected decimal ScaleFactor = 1;
+        readonly object _fillLock = new object();
+        const string buy = "buy";
+        const string sell = "sell";
+        /// <summary>
+        /// Currently limited to BTCUSD
+        /// </summary>
+        protected Symbol Symbol = Symbol.Create("BTCUSD", SecurityType.Forex, Market.Bitfinex);
+        /// <summary>
+        /// List of known orders
+        /// </summary>
+        public ConcurrentDictionary<int, Order> CachedOrderIDs = new ConcurrentDictionary<int, Order>();
+        /// <summary>
+        /// List of filled orders
+        /// </summary>
+        protected readonly FixedSizeHashQueue<int> FilledOrderIDs = new FixedSizeHashQueue<int>(10000);
+        /// <summary>
+        /// List of unknown orders
+        /// </summary>
+        protected readonly FixedSizeHashQueue<int> UnknownOrderIDs = new FixedSizeHashQueue<int>(1000);
+        /// <summary>
+        /// Name of wallet
+        /// </summary>
+        protected string Wallet;
+        const string _exchange = "bitfinex";
+        /// <summary>
+        /// Api Key
+        /// </summary>
+        protected string ApiKey;
+        /// <summary>
+        /// Api Secret
+        /// </summary>
+        protected string ApiSecret;
+        TradingApi.Bitfinex.BitfinexApi _restClient;
+        /// <summary>
+        /// Security Provider
+        /// </summary>
+        protected ISecurityProvider SecurityProvider;
+        #endregion
+
+        /// <summary>
+        /// Create bitfinex brokerage
+        /// </summary>
+        public BitfinexBrokerage(string apiKey, string apiSecret, string wallet, BitfinexApi restClient, decimal scaleFactor, ISecurityProvider securityProvider)
+            : base("bitfinex")
+        {
+            ApiKey = apiKey;
+            ApiSecret = apiSecret;
+            Wallet = wallet;
+            _restClient = restClient;
+            ScaleFactor = scaleFactor;
+            SecurityProvider = securityProvider;
+        }
+
+        /// <summary>
+        /// Determines if ticker polling is active
+        /// </summary>
+        public override bool IsConnected
+        {
+            get { return this._tickerToken != null && !this._tickerToken.IsCancellationRequested; }
+        }
+
+        private decimal GetPrice(Order order)
+        {
+            if (order is StopMarketOrder)
+            {
+                return ((StopMarketOrder)order).StopPrice * ScaleFactor;
```

Comment:
I notice the cfd support relies on something like a scale factor at security level so realize the question is larger than just this brokerage. The level of unit test coverage I've provided should be sufficient to instill confidence about making future changes.